db/src/db.rs

```diff
@@ -0,0 +1,541 @@
+        INSTEAD OF INSERT ON fulltext_values_view
+        WHEN EXISTS (SELECT 1 FROM fulltext_values WHERE text = new.text)
+        BEGIN
+          UPDATE fulltext_values SET searchid = new.searchid WHERE text = new.text;
+        END"#,
+   r#"CREATE TRIGGER insert_fulltext_searchid
+        INSTEAD OF INSERT ON fulltext_values_view
+        WHEN NOT EXISTS (SELECT 1 FROM fulltext_values WHERE text = new.text)
+        BEGIN
+          INSERT INTO fulltext_values (text, searchid) VALUES (new.text, new.searchid);
+        END"#,
+
+   // A view transparently interpolating fulltext indexed values into the datom structure.
+   r#"CREATE VIEW fulltext_datoms AS
+        SELECT e, a, fulltext_values.text AS v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value
+          FROM datoms, fulltext_values
+          WHERE datoms.index_fulltext IS NOT 0 AND datoms.v = fulltext_values.rowid"#,
+
+   // A view transparently interpolating all entities (fulltext and non-fulltext) into the datom structure.
+   r#"CREATE VIEW all_datoms AS
+        SELECT e, a, v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value
+          FROM datoms
+          WHERE index_fulltext IS 0
+        UNION ALL
+        SELECT e, a, v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value
+          FROM fulltext_datoms"#,
+
+   // Materialized views of the schema.
+   r#"CREATE TABLE idents (ident TEXT NOT NULL PRIMARY KEY, entid INTEGER UNIQUE NOT NULL)"#,
+   r#"CREATE TABLE schema (ident TEXT NOT NULL, attr TEXT NOT NULL, value BLOB NOT NULL, value_type_tag SMALLINT NOT NULL,
+        FOREIGN KEY (ident) REFERENCES idents (ident))"#,
+   r#"CREATE INDEX idx_schema_unique ON schema (ident, attr, value, value_type_tag)"#,
+   r#"CREATE TABLE parts (part TEXT NOT NULL PRIMARY KEY, start INTEGER NOT NULL, idx INTEGER NOT NULL)"#,
+   ];
+
+/// Additional SQL statements to be executed, in order, to create the Mentat SQL schema (version 2).
+/// We assume that the `V1_STATEMENTS` have been successfully executed (in order) before these are
+/// executed.
+#[cfg_attr(rustfmt, rustfmt_skip)]
+const V2_STATEMENTS: [&'static str; 0] = [];
+
+/// Set the SQLite user version.
+///
+/// Mentat manages its own SQL schema version using the user version.  See the [SQLite
+/// documentation](https://www.sqlite.org/pragma.html#pragma_user_version).
+fn set_user_version(conn: &rusqlite::Connection, version: i32) -> Result<()> {
+    conn.execute(&format!("PRAGMA user_version = {}", version), &[])
+        .chain_err(|| "Could not set_user_version")
+        .map(|_| ())
+}
+
+/// Get the SQLite user version.
+///
+/// Mentat manages its own SQL schema version using the user version.  See the [SQLite
+/// documentation](https://www.sqlite.org/pragma.html#pragma_user_version).
+fn get_user_version(conn: &rusqlite::Connection) -> Result<i32> {
+    conn.query_row("PRAGMA user_version", &[], |row| {
+        row.get(0)
```

Comment:
FWIW I _just_ fixed this in rusqlite ([PR](https://github.com/jgallagher/rusqlite/pull/210)); the `get` here will now panic instead of silently truncating. `get_checked` is also available if you want a `Result` instead.